### ISSUE TITLE
Compiler errors

### ISSUE THREAD

**Issue by lczech** [RosettaCommons/binder#13]
After I produced the bindings files, I now tried to compile everything, but there are a couple of problems. I am not entirely sure yet where they come from - my part, Pybind, or Binder. It seems, the latter, so I'm opening this issue.

#### Clang 3.9.1, WARN_FLAGS "-Wall -Wextra -Wno-unknown-pragmas -pedantic -pedantic-errors"
```
In file included from [...]/src/placement/function/functions_1.cpp:23:
In file included from [...]/pybind11/include/pybind11/pybind11.h:36:
In file included from [...]/pybind11/include/pybind11/attr.h:13:
[...]/pybind11/include/pybind11/cast.h:961:79: error: token pasting of ',' and __VA_ARGS__ is a GNU
      extension [-Werror,-Wgnu-zero-variadic-macro-arguments]
    struct always_construct_holder<holder_type> : always_construct_holder<void, ##__VA_ARGS__>  { }; \
                                                                              ^
[...]/pybind11/include/pybind11/cast.h:961:79: error: token pasting of ',' and __VA_ARGS__ is a GNU
      extension [-Werror,-Wgnu-zero-variadic-macro-arguments]
...
```
The first line `functions_1.cpp:23` is simply `#include <pybind11/pybind11.h>`.

The warning comes from a [GNU extension](https://gcc.gnu.org/onlinedocs/gcc-4.5.1/gcc/Variadic-Macros.html#Variadic-Macros). The interesting part is that all the includes (`attr.h`, `cast.h`) are part of the pybind header. So they should also happen when I try the minimal example of Pybind - but there, I don't get any errors. The minimal example works for me.

So, I tried to disable the warnings by leaving out the `pedantic` flag. But then, other errors occur:

#### Clang 3.9.1, WARN_FLAGS "-Wall -Wextra -Wno-unknown-pragmas"
```
In file included from [...]/src/placement/function/helper.cpp:1:
/usr/bin/../lib/gcc/x86_64-linux-gnu/4.9/../../../../include/c++/4.9/bits/hashtable_policy.h:34:15: error: expected '{'
namespace std _GLIBCXX_VISIBILITY(default)
              ^
/usr/bin/../lib/gcc/x86_64-linux-gnu/4.9/../../../../include/c++/4.9/bits/hashtable_policy.h:34:15: error: C++ requires a type
      specifier for all declarations
/usr/bin/../lib/gcc/x86_64-linux-gnu/4.9/../../../../include/c++/4.9/bits/hashtable_policy.h:34:35: error: expected expression
namespace std _GLIBCXX_VISIBILITY(default)
                                  ^
/usr/bin/../lib/gcc/x86_64-linux-gnu/4.9/../../../../include/c++/4.9/bits/hashtable_policy.h:34:43: error: expected ';' after top
      level declarator
namespace std _GLIBCXX_VISIBILITY(default)
                                          ^
...
```
Here, the first line `helper.cpp:1` is `#include <bits/hashtable_policy.h>`.

It seems a bit weird to me that Binder directly includes `bits` headers instead of using the "clean" names of those headers. But maybe this is needed. Anyway, does not seem to work, so I tried GCC instead. 

#### GNU 4.9.4, WARN_FLAGS "-Wall -Wextra -Wno-unknown-pragmas -pedantic -pedantic-errors"
```
[...]/src/placement/function/functions_1.cpp:27:52: error: ISO C99 requires rest arguments to be used
  PYBIND11_DECLARE_HOLDER_TYPE(T, std::shared_ptr<T>);
                                                    ^
[...]/src/placement/function/functions_1.cpp:28:36: error: ISO C99 requires rest arguments to be used
  PYBIND11_DECLARE_HOLDER_TYPE(T, T*);
                                    ^
[...]/src/placement/function/functions_1.cpp:27:53: error: extra ‘;’ [-Wpedantic]
  PYBIND11_DECLARE_HOLDER_TYPE(T, std::shared_ptr<T>);
                                                     ^
[...]/src/placement/function/functions_1.cpp:28:37: error: extra ‘;’ [-Wpedantic]
  PYBIND11_DECLARE_HOLDER_TYPE(T, T*);
                                     ^
...
```
The line that is causing this is the definition `PYBIND11_DECLARE_HOLDER_TYPE(T, std::shared_ptr<T>);`, as part of the `#ifndef BINDER_PYBIND11_TYPE_CASTER` block. So, this seems to be a similar problem as the first one with Clang. Next, naturally, I tried again without pedantic:
#### GNU 4.9.4, WARN_FLAGS "-Wall -Wextra -Wno-unknown-pragmas"
```
In file included from [...]/src/placement/function/helper.cpp:1:0:
/usr/include/c++/4.9/bits/hashtable_policy.h:34:15: error: expected ‘{’ before ‘_GLIBCXX_VISIBILITY’
 namespace std _GLIBCXX_VISIBILITY(default)
               ^
/usr/include/c++/4.9/bits/hashtable_policy.h:34:34: error: expected constructor, destructor, or type conversion before ‘(’ token
 namespace std _GLIBCXX_VISIBILITY(default)
                                  ^
In file included from /usr/include/c++/4.9/typeinfo:34:0,
                 from /usr/include/c++/4.9/debug/formatter.h:34,
                 from /usr/include/c++/4.9/debug/functions.h:41,
                 from /usr/include/c++/4.9/debug/debug.h:127,
                 from /usr/include/c++/4.9/bits/unique_ptr.h:34,
                 from [...]/src/placement/function/helper.cpp:2:
/usr/include/c++/4.9/exception:35:37: error: expected ‘}’ before end of line
 #pragma GCC visibility push(default)
                                     ^
...
```
Similar to the second Clang problem.

Any help is appreciated. What is going on here? Thanks in advance!

**Comment by lyskov** [comment]
Re `#include <bits/hashtable_policy.h>`: this is fixable but i will need a some info from you: could you please re-generate bindings with flag `--annotate-includes` then grep for `#include <bits/` and post results here?

**Comment by lyskov** [comment]
Re `Clang 3.9.1, WARN_FLAGS "-Wall -Wextra -Wno-unknown-pragmas -pedantic -pedantic-errors"`: this seems to be problem with Pybind11 so i guess submit it on main Pybind11 site: https://github.com/pybind/pybind11

Re `GNU 4.9.4, WARN_FLAGS "-Wall -Wextra -Wno-unknown-pragmas"`: in general it much easier to build results with Clang so unless you absolutely want GCC compatibility i suggest you to just use Clang (issues that you mention above should be easy to fix). Though this particular error is seems to the the same as the one that tripped clang, so lets address it first, when i have `grep` result from you i will look it up.

Thanks,

**Comment by lczech** [comment]
It is quite a long log file:
[bits_includes.log.zip](https://github.com/RosettaCommons/binder/files/804852/bits_includes.log.zip)

Okay, I'll wait for your feedback then before opening an issue at Pybind. Thanks!

**Comment by lyskov** [comment]
Perfect, this is exactly what i need, thank you!

**Comment by lyskov** [comment]
Ok, i just pushed 10cfde177403e436649f68af618bf116fec086b1, please see if it fix all of these issues. If it does not: please send me updated `grep` results.

**Comment by lyskov** [comment]
Re `make_pair("<bits/hashtable_policy.h>",   "<unordered_map>")` - great! I was not sure which include this should be because it seems possible to fit both `unordered_map` and `unordered_set`. I will make changes to upstream.

Re [1]: This does not make sense to me. Binder always uses namespaces for all its names (otherwise things will simply do not compile). So must be something else going on here. My guess is that most likely some of the headers pulled `std::string` into `::` namespace. Could you please check your headers for `using namespace` directive?

**Comment by lyskov** [comment]
Glad to hear that you was able to compile it!

Re other issues: One things i noticed in that grep results is that your projects seems to be using unique pointers for passing values around. This might get into conflict with Binder by default using `shared_ptr` as holder type. How do you want you Python code to behave? If you want it to follow the C++ convention of your library then it might be best to change default holder type `unique_ptr` instead...

**Comment by lczech** [comment]
Thanks!

Re `make_pair("<bits/hashtable_policy.h>", "<unordered_map>")`: Well, it is working in my case. Not sure if other users would need `unordered_set` instead...

Re [1]: No, I do not have any `using namespace std` directives, I always use the fully qualified `std::...` for everything.

Re `unique_ptr`: I have two uses cases for `unique_ptr` in my code:
 
 1. In order to allow inheritance, i.e., `std::vector< std::unique_ptr< BaseClass >>`. This might be changed to `shared_ptr`, but I'd like to avoid if necessary, in order to make clear that the `BaseClass` objects are tied to the lifetime of the `vector` (i.e., only live as long as the class that the `vector` is a member of).
 2. In order to guarantee stable memory addresses, again in a `std::vector< std::unique_ptr< MyClass >>`. I want to be able to change the contents of the vector, but have a stable address for the `MyClass` objects. A `std::list` would be an alternative here, but has other downsides, so I decided against it. Again, in this case, the objects should live as long as the container, so I used `unique_ptr` here, too.

In both cases, there is not much need for Python to directly work with the `unique_ptr` instances. I do not need to pass them around via my API - those parts that actually `expose` the `unique_ptrs` are meant for internal use in the library anyway. So, for the outside world (e.g., the Python interface), simple references are enough to work with my classes. Thus, I can simply exclude all the auto-generated bindings that concern `unique_ptrs`. Thanks!
